File: Day73/Part4/map_reveal.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>

using namespace std;

int solve_bfs(int start_r, int start_c, int R, int C, const vector<vector<int>>& grid) {
    priority_queue<tuple<int, int, int>, vector<tuple<int, int, int>>, greater<tuple<int, int, int>>> pq;
    vector<vector<bool>> visited(R, vector<bool>(C, false));
    
    long long current_power = grid[start_r][start_c];
    visited[start_r][start_c] = true;
    int count = 1;
    
    int dr[] = {-1, 1, 0, 0};
    int dc[] = {0, 0, -1, 1};
    
    // Add neighbors to PQ
    for(int i=0; i<4; ++i) {
        int nr = start_r + dr[i];
        int nc = start_c + dc[i];
        if(nr >= 0 && nr < R && nc >= 0 && nc < C) {
            pq.push({grid[nr][nc], nr, nc});
            visited[nr][nc] = true; // Mark as in-queue/visited to avoid duplicates
        }
    }
    
    while(!pq.empty()) {
        auto [val, r, c] = pq.top();
        
        if (current_power >= val) {
            pq.pop();
            current_power += val;
            count++;
            
            for(int i=0; i<4; ++i) {
                int nr = r + dr[i];
                int nc = c + dc[i];
                if(nr >= 0 && nr < R && nc >= 0 && nc < C && !visited[nr][nc]) {
                    visited[nr][nc] = true;
                    pq.push({grid[nr][nc], nr, nc});
                }
            }
        } else {
            // Cannot eat the smallest neighbor. Stuck.
            break;
        }
    }
    return count;
}

int main() {
    int R, C;
    if (cin >> R >> C) {
        vector<vector<int>> grid(R, vector<int>(C));
        int max_val = -1;
        for(int i=0; i<R; ++i) {
            for(int j=0; j<C; ++j) {
                cin >> grid[i][j];
                if (grid[i][j] > max_val) {
                    max_val = grid[i][j];
                }
            }
        }
        
        // Optimization: Only start from cells with max_val.
        // Starting with max_val gives the highest initial power.
        int max_revealed = 0;
        for(int i=0; i<R; ++i) {
            for(int j=0; j<C; ++j) {
                if (grid[i][j] == max_val) {
                    max_revealed = max(max_revealed, solve_bfs(i, j, R, C, grid));
                }
            }
        }
        
        cout << max_revealed << endl;
    }
    return 0;
}

File: Day73/Part4/map_reveal.py
import heapq

def solve_bfs(start_r, start_c, R, C, grid):
    pq = []
    visited = set()
    
    current_power = grid[start_r][start_c]
    visited.add((start_r, start_c))
    count = 1
    
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]
    
    # Add neighbors
    for i in range(4):
        nr, nc = start_r + dr[i], start_c + dc[i]
        if 0 <= nr < R and 0 <= nc < C:
            heapq.heappush(pq, (grid[nr][nc], nr, nc))
            visited.add((nr, nc))
            
    while pq:
        val, r, c = pq[0] # Peek
        
        if current_power >= val:
            heapq.heappop(pq)
            current_power += val
            count += 1
            
            for i in range(4):
                nr, nc = r + dr[i], c + dc[i]
                if 0 <= nr < R and 0 <= nc < C and (nr, nc) not in visited:
                    visited.add((nr, nc))
                    heapq.heappush(pq, (grid[nr][nc], nr, nc))
        else:
            break
            
    return count

def solve():
    try:
        line = input().split()
        if not line:
            return
        R, C = int(line[0]), int(line[1])
        
        grid = []
        for _ in range(R):
            grid.append([int(x) for x in input().split()])
            
        max_val = -1
        for r in range(R):
            for c in range(C):
                if grid[r][c] > max_val:
                    max_val = grid[r][c]
                    
        max_revealed = 0
        for r in range(R):
            for c in range(C):
                if grid[r][c] == max_val:
                    max_revealed = max(max_revealed, solve_bfs(r, c, R, C, grid))
                    
        print(max_revealed)
            
    except EOFError:
        pass

if __name__ == "__main__":
    solve()
